torch_geometric/tp.py

```diff
@@ -0,0 +1,172 @@
+                        # Otherwise just check type
+                        else:
+                            if not np.issubdtype(dtypes[colname],
+                                                 np.dtype(dt)):
+                                raise TypeError(
+                                    f"{dtypes[colname]} is not a subtype of {dt} for graph attribute {colname}"
+                                )
+
+        return f(*args, **kwargs)
+
+    return wrapper
+
+
+def _resolve_type(t: Any) -> Any:
+    """Adapted from dataenforce contib by @martijnentink:
+    https://github.com/CedricFR/dataenforce/pull/5"""
+    # support for NewType in type hinting
+    if isinstance(t, _AnnotatedType):
+        return t
+    if hasattr(t, "__supertype__"):
+        return _resolve_type(t.__supertype__)
+    # support for typing.List and typing.Dict
+    if hasattr(t, "__origin__"):
+        return _resolve_type(t.__origin__)
+    return t
+
+
+class DataMeta(GenericMeta):
+    """Metaclass for Data (internal)."""
+    def __new__(metacls, name, bases, namespace, **kargs):
+        return super().__new__(metacls, name, bases, namespace)
+
+    @_tp_cache
+    def __getitem__(self, parameters):
+        if hasattr(self, "__origin__") and (self.__origin__ is not None
+                                            or self._gorg is not Graph):
+            return super().__getitem__(parameters)
+        if parameters == ():
+            return super().__getitem__((_TypingEmpty, ))
+        if not isinstance(parameters, tuple):
+            parameters = (parameters, )
+        parameters = list(parameters)
+
+        check_only_specified = True
+        if parameters[-1] is ...:
+            check_only_specified = False
+            parameters.pop()
+
+        attributes, dtypes = _get_attribute_dtypes(parameters)
+
+        meta = DataMeta(self.__name__, self.__bases__, {})
+        meta.check_only_specified = check_only_specified
+        meta.attributes = attributes
+        meta.dtypes = dtypes
+        return meta
+
+
+def _get_attribute_dtypes(
+    p: Union[str, slice, list, set,
+             DataMeta]) -> Tuple[Set[str], Dict[str, Any]]:
+    attributes = set()
+    dtypes = {}
+    if isinstance(p, str):
+        attributes.add(p)
+    elif isinstance(p, slice):
+        attributes.add(p.start)
+        stop_type = _resolve_type(p.stop)
+        dtypes[p.start] = stop_type
+    elif isinstance(p, (list, set)):
+        for el in p:
+            subattributes, subdtypes = _get_attribute_dtypes(el)
+            attributes |= subattributes
+            dtypes.update(subdtypes)
+    elif isinstance(p, DataMeta):
+        attributes |= _get_attribute_dtypes(p)[0]
+        dtypes.update(_get_attribute_dtypes(p)[1])
+    else:
+        raise TypeError(
+            "Dataset[col1, col2, ...]: each col must be a string, list or set."
+        )
+    return attributes, dtypes
+
+
+class Graph(type(Data), extra=Generic, metaclass=DataMeta):
```

Comment:
It seems a bit unfortunate if the type name doesn't refer to `Data`. Maybe it should be `DataT` or something?

Related to this, will it work in an editor to have
```
g: Graph
g....
```
Will I get auto-completion on the functions defined on `Data` for `g`? I tried this pulling your PR and it didn't work but perhaps it was my IDE.

This works for example with numpy's generic NDArray type.

Reply:
I'm happy to discuss alternate names. I think reference to `Data` is sensible.

Wrt IDE autocompletion, I don't think this will work in this instance as we're inheriting from `type(Data)` rather than `Data`. I tried to inherit from `Data` directly but was running into some errors that I got stuck on. I agree it's not ideal. FWIW it's possible to use this implementation to inherit from `np.ndarray` (rather than `type(np.ndarray)` so there may be something in the `Data` implementation that's making this an issue.

Reply:
@Padarn I've renamed to `DataT` and `BatchT`.

I believe I've resolved the autocomplete issue. Let me know if it's still a problem in your IDE.

Reply:
Hey sorry for the slow reply been a bit busy: I just checked it out and it seems to be working. Thanks!